drivers/src/lms.rs

```diff
@@ -272,37 +289,156 @@ const LMOTS_P: [LmotsParameter; 9] = [
-        match algo_type {
-            LmsAlgorithmType::LmsSha256N32H5 => Ok((32, 5)),
-            LmsAlgorithmType::LmsSha256N32H10 => Ok((32, 10)),
-            LmsAlgorithmType::LmsSha256N32H15 => Ok((32, 15)),
-            LmsAlgorithmType::LmsSha256N32H20 => Ok((32, 20)),
-            LmsAlgorithmType::LmsSha256N32H25 => Ok((32, 25)),
-            LmsAlgorithmType::LmsSha256N24H5 => Ok((24, 5)),
-            LmsAlgorithmType::LmsSha256N24H10 => Ok((24, 10)),
-            LmsAlgorithmType::LmsSha256N24H15 => Ok((24, 15)),
-            LmsAlgorithmType::LmsSha256N24H20 => Ok((24, 20)),
-            LmsAlgorithmType::LmsSha256N24H25 => Ok((24, 25)),
-            LmsAlgorithmType::LmsReserved => {
-                raise_err!(InvalidLmsAlgorithmType)
-            }
+    }
+    raise_err!(InvalidLmotsAlgorithmType)
+}
+
+pub fn get_lms_parameters(algo_type: &LmsAlgorithmType) -> CaliptraResult<(u8, u8)> {
+    match algo_type {
+        LmsAlgorithmType::LmsSha256N32H5 => Ok((32, 5)),
+        LmsAlgorithmType::LmsSha256N32H10 => Ok((32, 10)),
+        LmsAlgorithmType::LmsSha256N32H15 => Ok((32, 15)),
+        LmsAlgorithmType::LmsSha256N32H20 => Ok((32, 20)),
+        LmsAlgorithmType::LmsSha256N32H25 => Ok((32, 25)),
+        LmsAlgorithmType::LmsSha256N24H5 => Ok((24, 5)),
+        LmsAlgorithmType::LmsSha256N24H10 => Ok((24, 10)),
+        LmsAlgorithmType::LmsSha256N24H15 => Ok((24, 15)),
+        LmsAlgorithmType::LmsSha256N24H20 => Ok((24, 20)),
+        LmsAlgorithmType::LmsSha256N24H25 => Ok((24, 25)),
+        LmsAlgorithmType::LmsReserved => {
+            raise_err!(InvalidLmsAlgorithmType)
+        }
+    }
+}
+
+pub fn parse_public_contents<const N: usize>(
+    public_string: &[u8],
+) -> CaliptraResult<LmsPublicKey<N>> {
+    if public_string.len() != (24 + N * 4) {
+        raise_err!(InvalidPublicKeyLength);
+    }
+    let mut pos = 0;
+    let lms_type = lookup_lms_algorithm_type(slice_to_num(&public_string[pos..pos + 4]))
+        .ok_or(err_u32!(InvalidLmsAlgorithmType))?;
+    pos += 4;
+
+    let lmots_type = lookup_lmots_algorithm_type(slice_to_num(&public_string[pos..pos + 4]))
+        .ok_or(err_u32!(InvalidLmotsAlgorithmType))?;
+    pos += 4;
+
+    let (hash_width, _) = get_lms_parameters(&lms_type)?;
+    if hash_width as usize != N * 4 {
+        raise_err!(InvalidHashWidth);
+    }
+
+    let lms_identifier: [u8; 16] = public_string[pos..][..16].try_into().unwrap();
+    pos += 16;
+
+    let mut temp = [0u32; N];
+
+    for t in temp.iter_mut() {
+        *t = slice_to_num(&public_string[pos..pos + 4]);
+        pos += 4;
+    }
+
+    let public_hash = HashValue::<N>::from(temp);
+
+    let pk = LmsPublicKey {
+        lms_type,
+        lmots_type,
+        lms_identifier,
+        root_hash: public_hash,
+    };
+    Ok(pk)
+}
+
+pub fn parse_signature_contents<const N: usize, const P: usize, const H: usize>(
+    signature: &[u8],
+) -> CaliptraResult<LmsSignature<N, P, H>> {
+    if signature.len() < 8 {
```

Comment:
The compiler isn't smart enough to realize that pos doesn't overflow below. Can you make this

```
    if signature.len() != 8 + N * 4 + (P * N * 4) + 4 + N * H {
        raise_err!(InvalidSignatureLength);
    }
```

Reply:
The optimizer really hates the function. I was able to get rid of the panic reference with this disgusting patch, but I recommend getting rid of the need to parse altogether; just have the sign function take the struct directly, and validate the few parameters that need to be validated there.

```
diff --git a/drivers/src/lms.rs b/drivers/src/lms.rs
index 4f4ace3e..9396655e 100644
--- a/drivers/src/lms.rs
+++ b/drivers/src/lms.rs
@@ -393,14 +393,18 @@ pub fn parse_signature_contents<const N: usize, const P: usize, const H: usize>(
 
     let mut y = [HashValue::<N>::default(); P];
     for t in y.iter_mut() {
-        let mut tmp = [0u32; N];
-        for tt in tmp.iter_mut().take(N) {
-            *tt = slice_to_num(&signature[pos..pos + 4]);
+        for tt in t.0.iter_mut() {
+            if pos > signature.len() - 4 {
+                raise_err!(InvalidSignatureDepth);
+            }
+            *tt = slice_to_num(&signature[pos..][..4]);
             pos += 4;
         }
-        *t = HashValue::<N>::from(tmp);
     }
-    let lms_type = lookup_lms_algorithm_type(slice_to_num(&signature[pos..pos + 4]))
+    if pos > signature.len() - 4 || pos + 4 > signature.len() {
+        raise_err!(InvalidSignatureLength);
+    }
+    let lms_type = lookup_lms_algorithm_type(slice_to_num(&signature[pos..][..4]))
         .ok_or(err_u32!(InvalidLmsAlgorithmType))?;
     pos += 4;
 
@@ -420,13 +424,15 @@ pub fn parse_signature_contents<const N: usize, const P: usize, const H: usize>(
 
     let mut path = [HashValue::<N>::default(); H];
     for t in path.iter_mut() {
-        let mut tmp = [0u32; N];
-        for tt in tmp.iter_mut().take(N) {
-            *tt = slice_to_num(&signature[pos..pos + 4]);
+        for tt in t.0.iter_mut() {
+            if pos > signature.len() - 4 {
+                raise_err!(InvalidSignatureDepth);
+            }
+            *tt = slice_to_num(&signature[pos..][..4]);
             pos += 4;
         }
-        *t = HashValue::<N>::from(tmp);
     }
+
```

Reply:
Done.